parser_utils.py

```diff
@@ -0,0 +1,331 @@
+        return "ArithmeticOp(func={}, exps={})".format(self.func, self.args)
+
+    def __add__(self, other):
+        return BinaryOperation("add", self, other)
+
+    def __radd__(self, other):
+        return BinaryOperation("add", other, self)
+
+    def __sub__(self, other):
+        return BinaryOperation("sub", self, other)
+
+    def __rsub__(self, other):
+        return BinaryOperation("sub", other, self)
+
+    def __mul__(self, other):
+        return BinaryOperation("mult", self, other)
+
+    def __rmul__(self, other):
+        return BinaryOperation("mult", other, self)
+
+    def __truediv__(self, other):
+        return BinaryOperation("div", self, other)
+
+    def __rtruediv__(self, other):
+        return BinaryOperation("div", other, self)
+
+    def __neg__(self):
+        return UnaryOperation("neg", self)
+
+    def __pow__(self, power):
+        return BinaryOperation("pow", self, power)
+
+    def __rpow__(self, power):
+        return BinaryOperation("pow", power, self)
+
+
+class UnaryOperation(ArithmeticOperation):
+    """Lightweight class for representing unary operations appearing within the grammar."""
+    def __init__(self, func, exp):
+        super().__init__(func, exp)
+
+    def __repr__(self):
+        if self.func == "neg":
+            return "-{}".format(self.args[0])
+
+        return "{}({})".format(*self.tuple)
+
+
+class BinaryOperation(ArithmeticOperation):
+    """Lightweight class for representing binary operations appearing within the grammar."""
+    def __init__(self, func, exp1, exp2):
+        super().__init__(func, exp1, exp2)
+
+    def __repr__(self):
+        if self.func == "add":
+            return "{} + {}".format(*self.args)
+
+        if self.func == "sub":
+            return "{} - {}".format(*self.args)
+
+        if self.func == "mult":
+            return "{} * {}".format(*self.args)
+
+        if self.func == "div":
+            return "{} / {}".format(*self.args)
+
+        if self.func == "pow":
+            return "{} ^ {}".format(*self.args)
+
+        return "{}({}, {})".format(*self.tuple)
+
+
+class Op:
+    """Class for representing quantum operations found in the grammar."""
+    def __init__(self, kind, name, params, wires):
+        self.kind = kind
+        self.name = name
+        self.params = params
+        self.wires = wires
+
+    def __repr__(self):
+        if not all(p is None for p in self.params):
+            return f"{self.name}({','.join(str(p) for p in self.params)}) {','.join(str(w) for w in self.wires)};"
+
+        return f"{self.name} {','.join(str(w) for w in self.wires)};"
+
+
+class Gate(Op):
+    """Class for representing unitary quantum gates."""
+    def __init__(self, name, params, wires):
+        super().__init__(Ops.GATE, name, params, wires)
+
+
+class Channel(Op):
+    """Class for representing non-unitary quantum channels."""
+    def __init__(self, name, params, wires):
+        super().__init__(Ops.CHANNEL, name, params, wires)
```

Comment:
There are no examples that use this, but it's trivial to add to the grammar (same syntax as gates